### PR TITLE
Support position reader of OSS, COS and S3

### PR DESCRIPTION
**Pull request by zhezhidashi**
### What changes are proposed in this pull request?

Support Position Reader of Aliyun OSS and Tencent COS, and Update position reader of AWS S3.
Add abstract class of ObjectPositionReader, and let S3APositionReader, OSSPositionReader and COSPositionReader extend it.

### Why are the changes needed?

The original code lacked support for OSS and COS position reader, and this PR can support them.

### Does this PR introduce any user facing changes?

No

### CODE REVIEW

File: dora/core/common/src/main/java/alluxio/PositionReader.java [hunk 1]
```diff
@@ -103,4 +104,26 @@ int readInternal(long position, ReadTargetBuffer buffer, int length)
...
+
+  /**
+   * Extract from read internal of different ufs position reader.
+   * read data from ufs
+   * @param in input stream
+   * @param buffer target byte buffer
+   * @param bytesToRead bytes to read
+   * @return bytes read, or -1 none of data is read
+   */
+  default int readDataInternal(InputStream in, ReadTargetBuffer buffer,
```
Comment (dbw9580) [inline dora/core/common/src/main/java/alluxio/PositionReader.java:116]: why is this a default impl in `PositionReader`? The method reads from a `InputStream`, has nothing to do with `PositionReader`.

Comment (dbw9580) [reply dora/core/common/src/main/java/alluxio/PositionReader.java:116]: I think what you need is something that is the equivalent of `MultiRangeObjectInputStream` but for `PositionReader`.

File: dora/underfs/oss/src/main/java/alluxio/underfs/oss/OSSPositionReader.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+      //      if start <= end < file length, read from start to end
+      //      if end >= file length, read from start to file length - 1
+      GetObjectRequest getObjectRequest = new GetObjectRequest(mBucketName, mPath);
+      getObjectRequest.setRange(position, position + bytesToRead - 1);
+      object = mClient.getObject(getObjectRequest);
+    } catch (OSSException e) {
+      String errorMessage = String
+          .format("Failed to open key: %s bucket: %s error: %s",
+              mPath, mBucketName, e.getMessage());
+      throw new IOException(errorMessage, e);
```
Comment (dbw9580) [inline dora/underfs/oss/src/main/java/alluxio/underfs/oss/OSSPositionReader.java:79]: The common logic is to open an input stream, and then read data out of it and feed it into the output buffer. So you can extract these into an abstract Position reader that takes in an input stream and adapts as a position reader.

**Review by dbw9580 — CHANGES_REQUESTED** [review]

File: dora/underfs/cos/src/main/java/alluxio/underfs/cos/COSPositionReader.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+  @Override
+  protected InputStream getObjectInputStream(
+      long position, ReadTargetBuffer buffer,
+      int bytesToRead, String errorMessage) throws IOException {
+    try {
+      GetObjectRequest getObjectRequest = new GetObjectRequest(mBucketName, mPath);
+      getObjectRequest.setRange(position, position + bytesToRead - 1);
+      mObject = mClient.getObject(getObjectRequest);
+    } catch (CosServiceException e) {
+      throw new IOException(errorMessage, e);
```
Comment (dbw9580) [inline dora/underfs/cos/src/main/java/alluxio/underfs/cos/COSPositionReader.java:64]: error message should be constructed in-place
```suggestion
      throw new IOException(String.format("Failed to get object %s in bucket %s", mPath, mBucketName, e);
```

File: dora/core/common/src/main/java/alluxio/underfs/ObjectPositionReader.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+  }
+
+  /**
+   * @param position position of the file to start reading data
+   * @param buffer target byte buffer
+   * @param bytesToRead bytes to read
+   * @param errorMessage error message to throw
+   * @return input stream of Object Storage's API
+   */
+  protected abstract InputStream getObjectInputStream(
```
Comment (dbw9580) [inline dora/core/common/src/main/java/alluxio/underfs/ObjectPositionReader.java:88]: use an actual verb to indicate that this is not a getter

```suggestion
  protected abstract InputStream openObjectInputStream(
```

File: dora/underfs/cos/src/main/java/alluxio/underfs/cos/COSPositionReader.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+/**
+ * Implementation of {@link ObjectPositionReader} that reads from COS object store.
+ */
+@ThreadSafe
+public class COSPositionReader extends ObjectPositionReader {
+
+  /**
+   * Client for operations with COS.
+   */
+  protected COSClient mClient;
```
Comment (dbw9580) [inline dora/underfs/cos/src/main/java/alluxio/underfs/cos/COSPositionReader.java:35]: can be final

File: dora/underfs/cos/src/main/java/alluxio/underfs/cos/COSPositionReader.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+
+  /**
+   * COS object.
+   */
+  protected COSObject mObject;
+
+  /**
+   * @param client             the Tencent COS client
+   * @param bucketNameInternal the bucket name
```
Comment (dbw9580) [inline dora/underfs/cos/src/main/java/alluxio/underfs/cos/COSPositionReader.java:44]: what's `bucketNameInternal`? if it's not the same thing as a bucket name, put the descriptions in the javadoc

**Review by dbw9580 — CHANGES_REQUESTED** [review]

**Review by dbw9580 — APPROVED** [review]
LGTM

**Comment by zhezhidashi** [comment]
@JiamingMai @LuQQiu Please review this PR. Thanks :)

**Comment by JiamingMai** [comment]
alluxio-bot, merge this please

**Pull request merged**